solutions/cses-1640.mdx

```diff
@@ -141,3 +123,135 @@ print("IMPOSSIBLE")
 </PySection>
 
 </LanguageSection>
+
+## Solution 2 - Two Pointers
+
+By keeping two pointers, one at each end of the list, we can greedily move the left and right pointers if the sum of the two elements are less than or greater than the target sum, finding the two indices in $\mathcal{O}(N)$ time.
+<LanguageSection>
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class TwoSum {
+	public static void main(String[] args) {
+		Kattio io = new Kattio();
+		long N, X;
+		N = io.nextInt();
+		X = io.nextInt();
+
+		List<Pair> a = new ArrayList<>();
+						  
+		// append the element and its index
+		for (int i = 0; i < N; i++) {
+			long x = io.nextInt();
+			a.add(new Pair(x, i + 1));
+		}
+		Collections.sort(a);
+		int i = 0, j = (int) N - 1;
+		while (i < j) {
+			// adjust the left and right pointers.
+			if (a.get(i).val + a.get(j).val > X)
+				j--;
+			else if (a.get(i).val + a.get(j).val < X)
+				i++;
+			else if (a.get(i).val + a.get(j).val == X) {
+				System.out.println(a.get(i).index + " " + a.get(j).index);
+				return;
+			}
+		}
+		io.println("IMPOSSIBLE");
+	}
+
+	static void sort(int[] a) {
+		ArrayList<Integer> l = new ArrayList<>();
+		for (int i : a)
+			l.add(i);
+		Collections.sort(l);
+		for (int i = 0; i < a.length; i++)
+			a[i] = l.get(i);
+	}
+	//CodeSnip{Kattio}
+
+	static class Pair implements Comparable<Pair> {
+		long val, index;
+
+		public Pair(long x, long y) {
+			this.val = x;
+	  		this.index = y;
+		}
+
+		public int compareTo(Pair o) {
+	  		return -Long.compare(o.val, val);
+		}
+	}
+}
+```
+</JavaSection>
+<CPPSection>
+
+```cpp
+//CodeSnip{CPP Short Template}
+
+int main() {
+	ll N, X;
+	cin >> N >> X;
+	vector<pair<int,int>> a;
+	
+	// append the element and its index
+	for (int i = 0; i < N; i++) {
+		ll x;
+		cin >> x;
+		a.pb({x, i + 1});
+	}
+	
+	sort(a.begin(), a.end());
+	int i = 0, j = N - 1;
+	
+	while (i < j) {
+		// adjust left and right pointers.
+		if(a[i].f + a[j].f > X)
+			j--;
+		else if (a[i].f + a[j].f < X)
+			i++;
+		else if (a[i].f + a[j].f == X) {
+			cout << a[i].s << " " << a[j].s;
+			return;
+		}
+	}
+	cout << "IMPOSSIBLE" << endl;
+}
+```
+</CPPSection>
+<PySection>
+
+```py
+nums = list(map(int, input().split()))
+a = []
+
+# append element and index.
+for i in range(n):
+	a.append([nums[i], i + 1])
+
+a.sort()
+ans = []
+left, right = 0, n - 1
+
+while left < right:
+	sumlr = a[left][0] + a[right][0]
+	if sumlr == x:
+		ans.append(str(a[right][1]))
+		ans.append(str(a[left][1]))
+		break
+	elif sumlr > x:
+		right -= 1
+	else:
+		left += 1
+
+# while loop terminated and couldn't find a pair.
```

Comment:
maybe change this comment to something like "print IMPOSSIBLE if we haven't found a pair"